scipy/spatial/src/distance_metrics.h

```diff
@@ -119,6 +119,152 @@ void transform_reduce_2d_(
     }
 }
 
+template <int ilp_factor=4,
+          typename T,
+          typename TransformFunc,
+          typename ProjectFunc = Identity,
+          typename ReduceFunc = Plus>
+void cosine_transform_reduce_2d_(
+    StridedView2D<T> out, StridedView2D<const T> x, StridedView2D<const T> y,
+    StridedView1D<T> xrownorms, StridedView1D<T> yrownorms, StridedView1D<T> ix,
+    const TransformFunc& map,
+    const ProjectFunc& project = Identity{},
+    const ReduceFunc& reduce = Plus{}) {
+    // Result type of calling map
+    using AccumulateType = typename std::decay<decltype(
+        map(std::declval<T>(), std::declval<T>()))>::type;
+    intptr_t xs = x.strides[1], ys = y.strides[1];
+
+    intptr_t i = 0;
+    if (xs == 1 && ys == 1) {
+        for (; i + (ilp_factor - 1) < x.shape[0]; i += ilp_factor) {
+            const T* x_rows[ilp_factor];
+            const T* y_rows[ilp_factor];
+            ForceUnroll<ilp_factor>{}([&](int k) {
+                x_rows[k] = &x(i + k, 0);
+                y_rows[k] = &y(i + k, 0);
+            });
+
+            AccumulateType dist[ilp_factor] = {};
+            for (intptr_t j = 0; j < x.shape[1]; ++j) {
+                ForceUnroll<ilp_factor>{}([&](int k) {
+                    auto val = map(x_rows[k][j], y_rows[k][j]);
+                    dist[k] = reduce(dist[k], val);
+                });
+            }
+
+            ForceUnroll<ilp_factor>{}([&](int k) {
+                out(i + k, 0) = 1.0 - project(dist[k])/(xrownorms(ix.shape[0]) * yrownorms(i + k));
```

Comment:
Instead of having completely new templates for `cosine`, `seclidean` and `jensenshannon`, what if we have just one new template that passes the indices `i` and `j` to `map`, and `i` to `project`. Then the metrics can use or ignore the indices as they require. If that performs the same then it would save a lot of duplication.